pinot-segment-local/src/test/java/org/apache/pinot/segment/local/segment/index/loader/ForwardIndexHandlerTest.java

```diff
@@ -469,18 +553,131 @@ public void testRewriteRawForwardIndexForMultipleColumns()
+    FieldSpec.DataType dataType = metadata.getDataType();
+    if (dataType == FieldSpec.DataType.STRING || dataType == FieldSpec.DataType.BYTES) {
+      // This value is based on the rows in createTestData().
+      dictionaryElementSize = 7;
+    } else if (dataType == FieldSpec.DataType.BIG_DECIMAL) {
+      dictionaryElementSize = 11;
+    }
+    validateMetadataProperties(col1, true, dictionaryElementSize, metadata.getCardinality(), metadata.getTotalDocs(),
+        dataType, metadata.getFieldType(), metadata.isSorted(), metadata.isSingleValue(),
+        metadata.getMaxNumberOfMultiValues(), metadata.getTotalNumberOfEntries(), metadata.isAutoGenerated(),
+        metadata.getMinValue(), metadata.getMaxValue());
+
+    // Col2 validation.
+    validateIndexMap(col2, true);
+    validateForwardIndex(col2, null);
+    // In column metadata, nothing other than hasDictionary and dictionaryElementSize should change.
+    dictionaryElementSize = 0;
+    metadata = existingSegmentMetadata.getColumnMetadataFor(col2);
+    dataType = metadata.getDataType();
+    if (dataType == FieldSpec.DataType.STRING || dataType == FieldSpec.DataType.BYTES) {
+      // This value is based on the rows in createTestData().
+      dictionaryElementSize = 7;
+    } else if (dataType == FieldSpec.DataType.BIG_DECIMAL) {
+      dictionaryElementSize = 11;
+    }
+    validateMetadataProperties(col2, true, dictionaryElementSize, metadata.getCardinality(), metadata.getTotalDocs(),
+        dataType, metadata.getFieldType(), metadata.isSorted(), metadata.isSingleValue(),
+        metadata.getMaxNumberOfMultiValues(), metadata.getTotalNumberOfEntries(), metadata.isAutoGenerated(),
+        metadata.getMinValue(), metadata.getMaxValue());
+  }
+
+  @Test(priority = 4)
+  public void testEnableDictionaryForSingleColumn()
+      throws Exception {
+    for (int i = 0; i < _noDictionaryColumns.size(); i++) {
+      SegmentMetadataImpl existingSegmentMetadata = new SegmentMetadataImpl(_segmentDirectory);
+      SegmentDirectory segmentLocalFSDirectory =
+          new SegmentLocalFSDirectory(_segmentDirectory, existingSegmentMetadata, ReadMode.mmap);
+      SegmentDirectory.Writer writer = segmentLocalFSDirectory.createWriter();
+
+      IndexLoadingConfig indexLoadingConfig = new IndexLoadingConfig(null, _tableConfig);
+      String column = _noDictionaryColumns.get(i);
+      indexLoadingConfig.getNoDictionaryColumns().remove(column);
+      ForwardIndexHandler fwdIndexHandler =
+          new ForwardIndexHandler(existingSegmentMetadata, indexLoadingConfig, _schema);
+      IndexCreatorProvider indexCreatorProvider = IndexingOverrides.getIndexCreatorProvider();
+      fwdIndexHandler.updateIndices(writer, indexCreatorProvider);
+
+      // Tear down before validation. Because columns.psf and index map cleanup happens at segmentDirectory.close()
+      segmentLocalFSDirectory.close();
+
+      validateIndexMap(column, true);
+      validateForwardIndex(column, null);
```

Comment:
Can we also check if ColumnIndexEntry exists / dictionary index buffer exists by calling `hasIndex()` ?